public/js/components/Editor.js

```diff
@@ -162,24 +170,49 @@ const Editor = React.createClass({
   },
 
   componentWillReceiveProps(nextProps) {
-    // Clear the currently highlighted line
-    if (isSourceForFrame(this.props.selectedSource, this.props.selectedFrame)) {
-      this.clearDebugLine(this.props.selectedFrame.location.line);
-    }
+    // This lifecycle method is responsible for updating the editor
+    // text.
+    const sourceText = nextProps.sourceText;
 
-    // Set the source text. The source text may not have been loaded
-    // yet. On startup, the source text may not exist yet.
-    if (nextProps.sourceText) {
-      this.setSourceText(nextProps.sourceText, this.props.sourceText);
+    if (!sourceText) {
+      this.setText("");
+      this.editor.setMode({ name: "text" });
+    } else if (!isTextForSource(sourceText)) {
+      // There are only 2 possible states: errored or loading. Do
+      // nothing except put a message in the editor.
+      this.setText(sourceText.get("error") || "Loading...");
+      this.editor.setMode({ name: "text" });
+    } else if (this.props.sourceText !== sourceText) {
+      // Only update it if the `sourceText` object has actually changed.
+      // It is immutable so it will always change when updated.
+      this.setText(sourceText.get("text"));
+      this.setMode(sourceText);
+      resizeBreakpointGutter(this.editor.codeMirror);
     }
+  },
 
-    if (this.props.selectedSource && !nextProps.selectedSource) {
-      this.editor.setText("");
+  componentDidUpdate(prevProps) {
+    // This is in `componentDidUpdate` so helper functions can expect
+    // `this.props` to be the current props. This lifecycle method is
+    // responsible for updating the editor annotations.
+    const { selectedLocation } = this.props;
+
+    // If the location is different and a new line is requested,
+    // update the pending jump line. Note that if jumping to a line in
+    // a source where the text hasn't been loaded yet, we will set the
+    // line here but not jump until rendering the actual source.
+    if (prevProps.selectedLocation !== selectedLocation &&
+       selectedLocation &&
+       selectedLocation.line != undefined) {
+      this.pendingJumpLine = selectedLocation.line;
     }
 
-    // Highlight the paused line if necessary
-    if (isSourceForFrame(nextProps.selectedSource, nextProps.selectedFrame)) {
-      this.setDebugLine(nextProps.selectedFrame.location.line);
+    // Only update and jump around in real source texts. This will
+    // keep the jump state around until the real source text is
+    // loaded.
+    if (this.props.sourceText && isTextForSource(this.props.sourceText)) {
+      this.updateDebugLine(prevProps, this.props);
+      this.highlightLine();
```

Comment:
You're right. There's a chance if we've selected a difference source, and its line is undefined, but we were in the middle of loading another source so we hadn't jumped yet, that it will inadvertently jump in the new source when it shouldn't. My new patch guards against this.